index.js

```diff
@@ -858,13 +868,161 @@ function normalizeQuery(inputQuery) {
+// function is used as an optimization to see whether we can avoid
+// augmenting the query to ignore deleted documents)
+function deletedDocCouldSatisfyQuery(query) {
+  // Any query with `{foo: value}` with non-null `value` will never
+  // match deleted documents (that are empty other than the `_type`
+  // field).
+  //
+  // This generalizes to additional classes of queries. Here’s a
+  // recursive description of queries that can't match a deleted doc:
+  // In general, a query with `{foo: X}` can't match a deleted doc
+  // if `X` is guaranteed to not match null or undefined. In addition
+  // to non-null values, the following clauses are guaranteed to not
+  // match null or undefined:
+  //
+  // * `{$in: [A, B, C]}}` where all of A, B, C are non-null.
+  // * `{$ne: null}`
+  // * `{$exists: true}`
+  // * `{$and: [A, B, C]}` where at least one of A, B, C is guaranteed to
+  //    not match null
+  // * `{$or: [A, B, C]}` where all of A, B, C are guaranteed to not match null
+  // * `{$gt: ...}`, `{gte: ...}`, `{$lt: ...}`, `{$lte: ...}`
+  //
+  // In addition, some queries that have `$and` or `$or` at the
+  // top-level can't match deleted docs:
+  // * `{$and: [A, B, C]}`, where at least one of A, B, C are queries
+  //   guaranteed to not match `{_type: null}`
+  // * `{$or: [A, B, C]}`, where all of A, B, C are queries guaranteed
+  //   to not match `{_type: null}`
+  //
+  // There are more queries that can't match deleted docs but they
+  // aren’t that common, e.g. ones using `$type` or bit-wise
+  // operators.
+  if (query.hasOwnProperty('$and')) {
+    if (Array.isArray(query.$and)) {
+      for (var i = 0; i < query.$and.length; i++) {
+        if (!deletedDocCouldSatisfyQuery(query.$and[i])) {
+          return false;
+        }
+      }
+      return true;
+    } else {
+      // Malformed? Play it safe.
+      return true;
+    }
+  }
+
+  if (query.hasOwnProperty('$or')) {
+    if (Array.isArray(query.$or)) {
+      for (var i = 0; i < query.$or.length; i++) {
+        if (deletedDocCouldSatisfyQuery(query.$or[i])) {
+          return true;
+        }
+      }
+      return false;
+    } else {
+      // Malformed? Play it safe.
+      return true;
+    }
+  }
+
+  for (var prop in query) {
+    // When using top-level operators that we don't understand, play
+    // it safe
+    if (prop[0] === '$')
+      return true;
+
+    if (!couldMatchNull(query[prop]))
+      return false;
+  }
+
+  return true;
+}
+
+function couldMatchNull(clause) {
+  if (typeof clause === "number" ||
+      typeof clause === "boolean" ||
+      typeof clause === "string") {
+    return false;
+  } else if (clause === null) {
+    return true;
+  } else if (isPojo(clause)) {
+    // Mongo interprets clauses with multiple properties with an
+    // implied 'and' relationship, e.g. {$gt: 3, $lt: 6}. If every
+    // part of the clause could match null then the full clause could
+    // match null.
+    for (var prop in clause) {
```

Comment:
rather than repeating the property to access the value in each of the cases below, i'd add a `var value = clause[property];` in this loop and use the value variable throughout the code below. Reduces opportunities for bugs from mistyping mistakes